experimental/torch_xla2/torch_xla2/train.py

```diff
@@ -0,0 +1,130 @@
+      if None, then no sharding is specified (i.e. for single device)
+  """
+  env = torch_xla2.default_env()
+  def loss(weights, buffers, args, label): # inputs are XLATensor
+    with env, jax.named_scope('compute_loss'):
+      if mark_fsdp_sharding_axis is not None:
+        args = mark_sharding(
+            args, 
+            jax.sharding.PartitionSpec(mark_fsdp_sharding_axis))
+      res = model_fn(weights, buffers, args)
+      if mark_fsdp_sharding_axis is not None:
+        res = mark_sharding(res, jax.sharding.PartitionSpec(mark_fsdp_sharding_axis))
+        label = mark_sharding(label, jax.sharding.PartitionSpec(mark_fsdp_sharding_axis))
+      l = loss_fn(res, label)
+      return l
+
+  loss = interop.gradient_checkpoint(loss, kwargs={'policy': remat_policy})
+  grad_fn = interop.jax_value_and_grad(loss)
+
+  def step(weights, buffers, opt_state, args, label): #inputs are array
+    with jax.named_scope('compute_gradient'):
+        loss, gradient = grad_fn(weights, buffers, args, label)
+
+    with jax.named_scope("optimizer_updates"):
+        updates, opt_state = interop.call_jax(
+            optax_optimizer.update,
+            gradient, opt_state, weights)
+        weights = interop.call_jax(optax.apply_updates, weights, updates)
+    return loss, weights, opt_state
+
+  return step
+
+  
+  
+  
+class Container:
+  pass
+
+class ScannedModule(torch.nn.Module):
+
+  def __init__(self, module_list, checkpoint_policy=None):
+    super().__init__()
+
+    self.c = None
+    assert module_list
+    self.c = Container()
+    self.c.one_mod = module_list[0]
+    self.checkpoint_policy = checkpoint_policy
+
+    weights = self._stack_layer_weights(module_list)
+    self.layer_weights_keys = list(self.c.one_mod.state_dict().keys())
+    self.params = torch.nn.ParameterDict({
+      self._param_name_new(k): v for k, v in weights.items()
+    })
+  
+  def _stack_layer_weights(self, module_list):
+    # Create weights such that, for every [n, m] weights
+    # becomes [k, n, m] where k is number of layer
+    # i.e. stacking layer weights together
+    temp = collections.defaultdict(list)
+    for m in module_list:
+      for k, v in m.state_dict().items():
+          temp[k].append(v)
+    res = {k: torch.stack(v) for k, v in temp.items()}
+    return res
+
+
+  def _param_name_new(self, old):
+      return '___'.join(old.split('.'))
+
+  def _param_name_old(self, new):
+      return '.'.join(new.split('___'))
+
+  def forward(self, *args, **kwargs):
+      assert not kwargs
+      weights = {k: self.params[self._param_name_new(k)] for k in self.layer_weights_keys}
+      scan = interop.torch_view(jax.lax.scan)
+
+      def eval_one_layer(args, weight):
+          # unpack args
+          h, *rest = args
+          newh = torch.func.functional_call(self.c.one_mod, weight, args)
+          # next layer's input; and residual to be added to list
+          return (newh, *rest), torch.ones(1, device='jax')
```

Comment:
Could the `torch.ones(1, ...)` simply be `None`? `None` is the base case of a PyTree and is what I managed to got working in `torch_xla`'s scan.

Reply:
done.